File: yuridekim/LeetCode/124_Binary_Tree_Maximum_Path_Sum.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def dfs(self, node):
        if not node: 
            return 0
        
        left_sum=self.dfs(node.left)
        if left_sum<0: left_sum=0
        right_sum=self.dfs(node.right)
        if right_sum<0: right_sum=0

        # max value of the path
        self.max_path = max(self.max_path, left_sum + right_sum + node.val)
        
        # add only one of the paths
        return max(left_sum, right_sum) + node.val
    def maxPathSum(self, root):
        self.max_path = float('-inf')
        self.dfs(root)
        return self.max_path

File: eileenjang/LeetCode/124_Binary_Tree_Maximum_Path_Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int sum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        calSum(root);
        return sum;
    }
    
    public int calSum(TreeNode root) {
        if (root == null) { return 0; }
        int left = Math.max(0, calSum(root.left));
        int right = Math.max(0, calSum(root.right));
        sum = Math.max(sum, root.val + left + right);
        return root.val + Math.max(left, right);
    }
}

// TC: 0(N), SC: O(N)
